### PR TITLE
GH-39914: [pyarrow] Reorder to_pandas extension dtype mapping

### PR DESCRIPTION
**Pull request by bretttully**
### Rationale for this change

This is a long standing [pandas ticket](https://github.com/pandas-dev/pandas/issues/53011) with some fairly horrible workarounds, where complex arrow types do not serialise well to pandas as the pandas metadata string is not parseable. However, `types_mapper` always had highest priority as it overrode what was set before. 

### What changes are included in this PR?

By switching the logical ordering, it means that we don't need to call `_pandas_api.pandas_dtype(dtype)` when using the pyarrow backend, thus resolving the issue of complex `dtype` with `list` or `struct`. It will likely still fail if the numpy backend is used, but at least this gives a working solution rather than an inability to load files at all.

### Are these changes tested?

Existing tests should stay unchanged and a new test for the complex type has been added

### Are there any user-facing changes?

**This PR contains a "Critical Fix".**
This makes `pd.read_parquet(..., dtype_backend="pyarrow")` work with complex data types where the metadata added by pyarrow during `pd.to_parquet` is not serialisable and currently throwing an exception. This issue currently prevents the use of pyarrow as the default backend for pandas.
* GitHub Issue: #39914

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> By switching the logical ordering, it means that we don't need to call `_pandas_api.pandas_dtype(dtype)` when using the pyarrow backend,

And because you added a `name not in ext_columns` to the subsequent methods to fill `ext_columns`, this should preserve the priority of the different methods to determine the pandas dtype? (metadata < pyarrow extension type < types_mapper)

File: python/pyarrow/tests/test_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_pandas.py:4436]: You might not need the roundtrip to parquet, but a `table = pa.table(df); result = table.to_pandas(types_mapper=pd.ArrowDtype)` should be sufficient to test this?

I know this doesn't test exactly `pd.read_parquet` in its entirety, but it should test the relevant part on the pyarrow side, and an actual pd.read_parquet test can still be added to pandas

**Comment by bretttully** [comment]
> > By switching the logical ordering, it means that we don't need to call `_pandas_api.pandas_dtype(dtype)` when using the pyarrow backend,
> 
> And because you added a `name not in ext_columns` to the subsequent methods to fill `ext_columns`, this should preserve the priority of the different methods to determine the pandas dtype? (metadata < pyarrow extension type < types_mapper)

Yes, exactly. Priority remains the same, but functions are skipped if the field already has a type, meaning that the code causing the error is no longer called if types_mapper is provided.

File: python/pyarrow/tests/test_pandas.py
Comment (bretttully) [reply python/pyarrow/tests/test_pandas.py:4436]: The error only gets thrown once the pandas metadata is added to the table. That's why I have used a round-trip test. Is there another way to generate that metadata and set it on the table before calling to_pandas?

**Comment by jorisvandenbossche** [comment]
The `test_dlpack` failure in the tests you can ignore (https://github.com/apache/arrow/issues/44728)

File: python/pyarrow/tests/test_pandas.py
Comment (bretttully) [reply python/pyarrow/tests/test_pandas.py]: Addressed in https://github.com/apache/arrow/pull/44720/commits/1c23076fb37f8903762c7416db12e8e453e23366

File: python/pyarrow/tests/test_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_pandas.py:4415]:
```suggestion
def test_to_pandas_extension_dtypes_mapping_complex_type():
    # https://github.com/apache/arrow/pull/44720
```

**Review by jorisvandenbossche — APPROVED** [review]
Looks good! 
I triggered CI again

**Comment by bretttully** [comment]
Thanks @jorisvandenbossche -- is the process that I can merge this following approval, or is that done by a core maintainer?

**Comment by raulcd** [comment]
> is the process that I can merge this following approval, or is that done by a core maintainer?

A committer will merge, probably @jorisvandenbossche in this specific case, once everything is running and addressed. I've triggered CI for the latest changes.

File: python/pyarrow/tests/test_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_pandas.py:4414]:
```suggestion
```

Small whitespace issue the linter is complaining about

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit -g python

**Comment by jorisvandenbossche** [comment]
@raulcd it seems something is going wrong with the minimal test builds (eg example-python-minimal-build-fedora-conda). The logs indicate "Successfully installed pyarrow-0.1.dev16896+ge3b9892", which then messes up pandas detection of the pyarrow version (for the pyarrow integration in pandas, pandas checks if pyarrow is recent enough and otherwise errors), giving some test failures. 

(but also not entirely sure how this PR causes this issue, since I don't see the nightlies fail for the minimal builds at the moment)

**Comment by jorisvandenbossche** [comment]
(the other failures are the known nightly dlpack failures)

**Comment by raulcd** [comment]
> The logs indicate "Successfully installed pyarrow-0.1.dev16896+ge3b9892"

From the git checkout I see is pulling from the remote on ` Syncing repository: bretttully/arrow`. I recall an issue if dev tags are not present we are unable to detect the correct version. The remote doesn't seem to have other branches and/or tags.

**Comment by raulcd** [comment]
I've opened an issue because we should find a way to not fail if the dev tag is not present:
- https://github.com/apache/arrow/issues/44803

**Comment by jorisvandenbossche** [comment]
Thanks for investigating that! 

So then to resolve this here, @bretttully should fetch the upstream tags and push that to his fork? Something like
```
git fetch upstream
git push origin --tags
```

(assuming upstream is apache/arrow and origin is bretttully/arrow)

**Comment by bretttully** [comment]
I have merged `upstream/main` and pushed tags. Let's see if this works...

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-*

**Comment by bretttully** [comment]
Is there anything else for me to do here?

**Comment by raulcd** [comment]
> Is there anything else for me to do here?

I don't think so. I am not comfortable with this area of our codebase so I'll let @jorisvandenbossche merge once he's happy about it, but as he already approved, he might do that soon.

**Comment by jorisvandenbossche** [comment]
> Is there anything else for me to do here?

No, just me getting back to merge it!

**Comment by bretttully** [comment]
🚀  Thanks for all your help here @jorisvandenbossche!

**Pull request merged**
